### PR TITLE
Skip processing snapshots of type Overwrite during readStream

### PR DESCRIPTION
**Pull request by SreeramGarlapati**
Delete operations on iceberg table can translate to Snapshots of type `DataOperations.DELETE` or `DataOperations.OVERWRITE`.

`DataOperations.DELETE` - when the delete operation on the table translates to full file deletes (for ex: deletes wiping out partitions using partitionKey level predicates)

`DataOperations.OVERWRITE` - is the most common case of Delete.

@rdblue / @aokolnychyi / @RussellSpitzer / @kbendick - pl. lemme know what you folks think & once we achieve principle level alignment - I will add unittests.

### CODE REVIEW

**Comment by SreeramGarlapati** [comment]
https://github.com/apache/iceberg/projects/2

**Comment by SreeramGarlapati** [comment]
I stumbled on the exact same change made by @kbendick : https://github.com/apache/iceberg/pull/3267 
@kbendick - truly apologize - I did not get a chance to look at this (I had to turn off my git notifications at some point in the past). Truly appreciate - if you can reopen your branch and get that going. Or - I can merge your changes into this PR. Either way works.

**Comment by tprelle** [comment]
Hi @SreeramGarlapati @kbendick,
I have some concern to link Overwrite and Delete because in table v1 so Copy on Write mode. A merge with an update will create a overwrite snapshot. So if you skip overwrite snapshot you will miss all theses changes.
For example delta.io allow user to choose to resend everything if it's copy on write mode.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java [hunk 1]
```diff
@@ -205,6 +205,11 @@ private boolean shouldProcess(Snapshot snapshot) {
             "Cannot process delete snapshot : %s. Set read option %s to allow skipping snapshots of type delete",
             snapshot.snapshotId(), SparkReadOptions.STREAMING_SKIP_DELETE_SNAPSHOTS);
         return false;
+      case DataOperations.OVERWRITE:
```
Comment (rdblue) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:208]: I don't think that this should use the same configuration to skip deletes and overwrites. Overwrites are different and I think that we should at a minimum have a different property. I would also prefer to have some additional clarity on how we plan to eventually handle this. We could skip overwrites, but what about use cases where they are probably upserts? What about when they're created by copy-on-write MERGE operations?

Comment (kbendick) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:208]: +1. I would prefer they be two separate configs, but also that we have a plan for the longer term to handle sending out these row deltas.

I'd be ok with getting a PR in to ignore `OVERWRITE`, but this isn't something we should ignore in the longer term (or even really the near-to-medium term) as others have mentioned.

Personally I would consider using a schema similar to the delta.io change capture feed that has a dataframe with the before image / after image (row before and after update) and then the type of operation for each row (insert, delete, update_before, update_after).

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java [hunk 2]
```diff
@@ -205,6 +205,11 @@ private boolean shouldProcess(Snapshot snapshot) {
...
+        Preconditions.checkState(skipDelete,
+            "Cannot process overwrite snapshot : %s. Set read option %s to allow skipping snapshots of type overwrite",
+            snapshot.snapshotId(), SparkReadOptions.STREAMING_SKIP_DELETE_SNAPSHOTS);
+        return false;
       default:
```
Comment (RussellSpitzer) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:213]: Should add a test for this new conf option

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java (hunk 1 above)
Comment (rajarshisarkar) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:210]: I have added a separate config to skip overwrites.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java (hunk 2 above)
Comment (rajarshisarkar) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:215]: I have added the Unit Test.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java (hunk 1 above)
Comment (rajarshisarkar) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:210]: I will discuss with @SreeramGarlapati and will update on the plan to eventually handle upserts.

Comment (rajarshisarkar) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:210]: I connected with @SreeramGarlapati to contribute on this PR.

Comment (rajarshisarkar) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:210]: I have added a separate config to skip overwrites. I will discuss with @SreeramGarlapati and will update on the plan to eventually handle upserts.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java (hunk 2 above)
Comment (rajarshisarkar) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:215]: I have added the Unit Test.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java (hunk 1 above)
Comment (SreeramGarlapati) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:216]: I agree with @kbendick - we will invest in emitting change feed - similar to delta format.

**Comment by rajarshisarkar** [comment]
Hi @rdblue @kbendick @RussellSpitzer Request you to please review.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java [hunk 3]
```diff
@@ -200,13 +202,25 @@ public void stop() {
...
-        "Cannot process %s snapshot: %s", op.toLowerCase(Locale.ROOT), snapshot.snapshotId());
-    return op.equals(DataOperations.APPEND);
+    switch (op) {
+      case DataOperations.APPEND:
+        return true;
+      case DataOperations.REPLACE:
+        return false;
+      case DataOperations.DELETE:
+        Preconditions.checkState(skipDelete,
+            "Cannot process delete snapshot : %s. Set read option %s to allow skipping snapshots of type delete",
```
Comment (rdblue) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:212]: I don't see a need to change the error message here. It was short and to the point before. Can you revert the change to it?

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java [hunk 4]
```diff
@@ -200,13 +202,25 @@ public void stop() {
...
+            snapshot.snapshotId(), SparkReadOptions.STREAMING_SKIP_DELETE_SNAPSHOTS);
+        return false;
+      case DataOperations.OVERWRITE:
+        Preconditions.checkState(skipOverwrite,
+            "Cannot process overwrite snapshot : %s. Set read option %s to allow skipping snapshots of type overwrite",
```
Comment (rdblue) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:217]: Can you copy the delete error message, but change "delete" to "overwrite"?

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java [hunk 5]
```diff
@@ -200,13 +202,25 @@ public void stop() {
...
+            snapshot.snapshotId(), SparkReadOptions.STREAMING_SKIP_OVERWRITE_SNAPSHOTS);
+        return false;
+      default:
+        throw new IllegalStateException(String.format(
+            "Cannot process %s snapshot: %s", op.toLowerCase(Locale.ROOT), snapshot.snapshotId()));
```
Comment (rdblue) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:222]: How about "Cannot process unknown snapshot operation: %s (snapshot id %s)"?

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java [hunk 6]
```diff
@@ -481,6 +481,35 @@ public void testReadStreamWithSnapshotTypeDeleteAndSkipDeleteOption() throws Exc
...
   }
 
+  @Test
+  public void testReadStreamWithSnapshotTypeDeleteAndSkipOverwriteOption() throws Exception {
+    table.updateSpec()
+        .removeField("id_bucket")
+        .addField(ref("id"))
+        .commit();
+
+    table.refresh();
```
Comment (rdblue) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java:491]: No need for refresh calls after altering a table. The table instance that was just committed will be up to date afterwards.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java [hunk 7]
```diff
@@ -481,6 +481,35 @@ public void testReadStreamWithSnapshotTypeDeleteAndSkipDeleteOption() throws Exc
...
+
+    // fill table with some data
+    List<List<SimpleRecord>> dataAcrossSnapshots = TEST_DATA_MULTIPLE_SNAPSHOTS;
+    appendDataAsMultipleSnapshots(dataAcrossSnapshots);
+
+    table.refresh();
```
Comment (rdblue) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java:497]: Also unnecessary. The table will be refreshed as part of the overwrite commit.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java (hunk 6 above)
Comment (rajarshisarkar) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java]: I have removed the refresh call.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java (hunk 7 above)
Comment (rajarshisarkar) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java]: I have removed the refresh call.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java (hunk 5 above)
Comment (rajarshisarkar) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java]: Thanks, this message seems better.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java (hunk 4 above)
Comment (rajarshisarkar) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java]: Sure.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java (hunk 3 above)
Comment (rajarshisarkar) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java]: Sure.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java [hunk 8]
```diff
@@ -200,13 +202,25 @@ public void stop() {
...
-        "Cannot process %s snapshot: %s", op.toLowerCase(Locale.ROOT), snapshot.snapshotId());
-    return op.equals(DataOperations.APPEND);
+    switch (op) {
+      case DataOperations.APPEND:
+        return true;
+      case DataOperations.REPLACE:
+        return false;
+      case DataOperations.DELETE:
+        Preconditions.checkState(skipDelete,
+                "Cannot process delete snapshot: %s, to ignore deletes, set %s=true.",
```
Comment (rdblue) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:212]: Indentation isn't correct.

Comment (rdblue) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java]: Fixed.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Merged. Thanks @SreeramGarlapati and @rajarshisarkar!
